aws/data_source_aws_region.go

```diff
@@ -36,48 +35,74 @@ func dataSourceAwsRegion() *schema.Resource {
 }
 
 func dataSourceAwsRegionRead(d *schema.ResourceData, meta interface{}) error {
-	conn := meta.(*AWSClient).ec2conn
 	currentRegion := meta.(*AWSClient).region
 
-	req := &ec2.DescribeRegionsInput{}
+	var matchingRegion *endpoints.Region
 
-	req.RegionNames = make([]*string, 0, 2)
-	if name := d.Get("name").(string); name != "" {
-		req.RegionNames = append(req.RegionNames, aws.String(name))
+	if v, ok := d.GetOk("endpoint"); ok {
+		endpoint := v.(string)
+		for _, partition := range endpoints.DefaultPartitions() {
+			for _, region := range partition.Regions() {
+				regionEndpointEc2, err := region.ResolveEndpoint(endpoints.Ec2ServiceID)
+				if err != nil {
+					return err
+				}
+				if strings.TrimPrefix(regionEndpointEc2.URL, "https://") == endpoint {
+					matchingRegion = &region
+					break
+				}
+			}
+		}
+		if matchingRegion == nil {
+			return fmt.Errorf("region not found for endpoint: %s", endpoint)
+		}
 	}
 
-	if d.Get("current").(bool) {
-		req.RegionNames = append(req.RegionNames, aws.String(currentRegion))
+	if v, ok := d.GetOk("name"); ok {
+		name := v.(string)
+		for _, partition := range endpoints.DefaultPartitions() {
+			for _, region := range partition.Regions() {
+				if region.ID() == name {
+					if matchingRegion != nil && (*matchingRegion).ID() != name {
+						return fmt.Errorf("multiple regions matched; use additional constraints to reduce matches to a single region")
+					}
+					matchingRegion = &region
+					break
+				}
+			}
+		}
+		if matchingRegion == nil {
+			return fmt.Errorf("region not found for name: %s", name)
+		}
```

Comment:
Likewise - how do you feel about extracting this into `matchingRegion, err := findRegionByName(name)`?

Reply:
👍 👍